src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/ExtractToCssCodeActionResolver.cs

```diff
@@ -0,0 +1,157 @@
+
+        // VS Code in Windows expects path to start with '/'
+        cssFilePath = _languageServerFeatureOptions.ReturnCodeActionAndRenamePathsWithPrefixedSlash && !cssFilePath.StartsWith("/")
+            ? '/' + cssFilePath
+            : cssFilePath;
+
+        var cssFileUri = LspFactory.CreateFilePathUri(cssFilePath);
+
+        var text = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+
+        var cssContent = text.GetSubTextString(new TextSpan(actionParams.ExtractStart, actionParams.ExtractEnd - actionParams.ExtractStart)).Trim();
+        var removeRange = codeDocument.Source.Text.GetRange(actionParams.RemoveStart, actionParams.RemoveEnd);
+
+        var codeDocumentIdentifier = new OptionalVersionedTextDocumentIdentifier { DocumentUri = new(documentContext.Uri) };
+        var cssDocumentIdentifier = new OptionalVersionedTextDocumentIdentifier { DocumentUri = new(cssFileUri) };
+
+        using var changes = new PooledArrayBuilder<SumType<TextDocumentEdit, CreateFile, RenameFile, DeleteFile>>(capacity: 3);
+
+        // First, an edit to remove the script tag and its contents.
+        changes.Add(new TextDocumentEdit
+        {
+            TextDocument = codeDocumentIdentifier,
+            Edits = [LspFactory.CreateTextEdit(removeRange, string.Empty)]
+        });
+
+        if (_fileSystem.FileExists(cssFilePath))
+        {
+            // CSS file already exists, insert the content at the end.
+            GetLastLineNumberAndLength(cssFilePath, out var lastLineNumber, out var lastLineLength);
+
+            changes.Add(new TextDocumentEdit
+            {
+                TextDocument = cssDocumentIdentifier,
+                Edits = [LspFactory.CreateTextEdit(
+                    position: (lastLineNumber, lastLineLength),
+                    newText: lastLineNumber == 0
+                        ? cssContent
+                        : Environment.NewLine + Environment.NewLine + cssContent)]
+            });
+        }
+        else
+        {
+            // No CSS file, create it and fill it in
+            changes.Add(new CreateFile { DocumentUri = cssDocumentIdentifier.DocumentUri });
+            changes.Add(new TextDocumentEdit
+            {
+                TextDocument = cssDocumentIdentifier,
+                Edits = [LspFactory.CreateTextEdit(position: (0, 0), cssContent)]
+            });
+        }
+
+        return new WorkspaceEdit
+        {
+            DocumentChanges = changes.ToArray(),
+        };
+    }
+
+    private void GetLastLineNumberAndLength(string cssFilePath, out int lastLineNumber, out int lastLineLength)
+    {
+        using var stream = _fileSystem.ReadStream(cssFilePath);
+        GetLastLineNumberAndLength(stream, bufferSize: 4096, out lastLineNumber, out lastLineLength);
+    }
+
+    private static void GetLastLineNumberAndLength(Stream stream, int bufferSize, out int lastLineNumber, out int lastLineLength)
```

Comment:
>GetLastLineNumberAndLength [](http://example.com/codeflow?start=24&length=26)

nit: Seems like this could go to StreamHelpers.cs or StreamExtensions.cs somewhere, and then you wouldn't need TestAccessor

Reply:
It could, but also feels pretty niche to me so I'm not sure